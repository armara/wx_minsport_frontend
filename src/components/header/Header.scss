@import "styles/mixins";
@import "styles/colors";
@import "styles/icons";

@mixin overflow() {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.app-header {
  position: relative;
  box-shadow: 0 -17px 20px 0 $black;
  background-color: $white;
  width: 100%;
  height: 60px;
  margin-bottom: 20px;
  padding: 0 30px 0 20px;
  display: flex;
  align-items: center;
  justify-content: flex-end;

  nav {
    display: flex;
    flex-grow: 1;
    align-items: center;
    justify-content: center;
  }

  .header_profile-block {
    position: absolute;
    top: 8px;
    right: 30px;
    display: flex;
    align-items: center;
    justify-content: space-between;
    flex-direction: column;

    .header_workplace-title {
      font-weight: 700;
      max-width: 400px;
      height: 45px;
      align-self: flex-end;
      justify-content: flex-start;

      div {
        @include overflow();
      }
    }

    .header_where-user-works {
      margin-top: 5px;
      border: 1px solid $light-gray;
      border-radius: 6px;
      box-shadow: 0 1px 28px -21px $davy-gray;
      background-color: $white;
      min-width: 200px;
      max-width: 400px;
      z-index: 5;

      &.hidden {
        display: none;
      }

      .header_where-user-works_personal {
        border-bottom: 1px solid $light-gray;
        padding: 5px 12px;

        div {
          @include overflow();
        }

        div:first-child {
          font-weight: 700;
        }

        div + div {
          font-size: 12px;
        }
      }

      .base-button {
        outline: none;
        border: none;
        width: 100%;
        border-radius: 0;
        background-color: transparent;
      }

      li {
        &:hover {
          background-color: $light-gray;
        }

        .base-button {
          display: flex;
          flex-flow: row nowrap;
          align-items: center;
          justify-content: flex-start;
          white-space: pre-wrap;
          height: auto;

          div {
            text-align: left;
            line-height: 20px;
          }
        }
      }

      li + li {
        border-top: 1px solid $light-gray;
      }

      .logout-btn {
        border-top: 1px solid $light-gray;
        font-weight: 700;

        &:hover {
          background-color: $light-gray;
          margin-bottom: 0;
        }
      }
    }

    .base-button {
      i {
        margin-right: 10px;
      }
    }
  }
}