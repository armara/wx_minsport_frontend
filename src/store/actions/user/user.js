import {
  SET_FACILITY,
  SET_IS_LOADING,
  SET_ERROR,
  SET_AREAS,
  SET_ZONES,
  SELECT_AREA_ID,
  SET_ALL_ZONES,
  SET_ALL_FACILITIES,
  SET_WHERE_USER_WORKS,
} from 'store/types/user'

export const setFacility = payload => ({
  type: SET_FACILITY,
  payload,
})

export const setAreas = payload => ({
  type: SET_AREAS,
  payload,
})

export const setIsLoading = payload => ({
  type: SET_IS_LOADING,
  payload,
})

export const setError = error => ({
  type: SET_ERROR,
  payload: error,
})

export const setAreaId = areaId => ({
  type: SELECT_AREA_ID,
  payload: areaId,
})

export const setZones = payload => ({
  type: SET_ZONES,
  payload,
})

export const setAllZones = payload => ({
  type: SET_ALL_ZONES,
  payload,
})

export const setAllFacilities = payload => ({
  type: SET_ALL_FACILITIES,
  payload,
})

export const setWhereUserWorks = payload => ({
  type: SET_WHERE_USER_WORKS,
  payload,
})
