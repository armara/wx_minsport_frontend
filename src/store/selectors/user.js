import { createSelector } from 'reselect'

export const userSelector = state => state.user

export const userFacilitySelector = createSelector(
  userSelector,
  user => user.facility || {}
)

export const userFacilitySubdistrictIdSelector = createSelector(
  userFacilitySelector,
  facility => facility.subdistrictId
)

export const userAreasSelector = createSelector(
  userSelector,
  user => user.areas
)

export const userAreasSelectorWithTitleKey = createSelector(
  userAreasSelector,
  areas => areas.map(({ name, ...rest }) => ({ ...rest, title: name }))
)

export const userZones = createSelector(
  userSelector,
  user => user.zones
)

export const userZonesWithTitleKey = createSelector(
  userZones,
  zones => zones.map(({ name, ...rest }) => ({ ...rest, title: name }))
)

export const userAllFacilitiesSelector = createSelector(
  userSelector,
  user => user.allFacilities
)

export const userAllZones = createSelector(
  userSelector,
  user => user.allZones
)

export const userSelectedAreaId = createSelector(
  userSelector,
  user => user.selectedAreaId
)

export const selectWhereUserWorks = createSelector(
  userSelector,
  user => user.whereUserWorks
)
