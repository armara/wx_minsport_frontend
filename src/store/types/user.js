const PREFIX = 'FACILITY'

export const SET_FACILITY = `${PREFIX}/SET_FACILITY`
export const SET_AREAS = `${PREFIX}/SET_AREAS`
export const SET_IS_LOADING = `${PREFIX}/SET_IS_LOADING`
export const SET_ERROR = `${PREFIX}/SET_ERROR`
export const SET_ZONES = `${PREFIX}/SET_ZONES`
export const SELECT_AREA_ID = `${PREFIX}/SELECT_AREA_ID`
export const SET_ALL_ZONES = `${PREFIX}/SET_ALL_ZONES`
export const SET_ALL_FACILITIES = `${PREFIX}/SET_ALL_FACILITIES`
export const SET_WHERE_USER_WORKS = `${PREFIX}/SET_WHERE_USER_WORKS`
