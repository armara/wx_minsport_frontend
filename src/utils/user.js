import pick from 'lodash.pick'

export const updateUserData = (currentUserData, nextUserData = {}) => {
  localStorage.setItem(
    'user',
    JSON.stringify({
      ...currentUserData,
      ...nextUserData,
    })
  )
}

export const createWorkplace = (type, id) => ({ type, id })

export const getUserData = () => {
  return JSON.parse(localStorage.getItem('user')) || {}
}

export const getUserWorkplace = () => getUserData().workPlace || {}
export const getUserWorkplaceType = () => getUserWorkplace().type || ''
export const getUserWorkplaceId = () => getUserWorkplace().id || ''
const workFields = ['facilities', 'organizations', 'federations', 'ministries']

export const getWhereUserWorks = () => pick(getUserData(), workFields)
const personalFields = ['firstName', 'surName', 'patronymicName', 'email']
export const getUserPersonalFields = () => pick(getUserData(), personalFields)
export const getUserName = () => {
  const {
    firstName = '',
    surName = '',
    patronymicName = '',
  } = getUserPersonalFields()

  return `${surName} ${(firstName[0] || '').toUpperCase()}${
    firstName ? '.' : ''
  } ${(patronymicName[0] || '').toUpperCase()}${
    patronymicName ? '.' : ''
  }`.trim()
}
export const getFullUserName = () => {
  const {
    firstName = '',
    surName = '',
    patronymicName = '',
  } = getUserPersonalFields()

  return `${surName} ${firstName} ${patronymicName}`.trim()
}

export const changeUserWorkplace = (type, id) => {
  const currentUserData = getUserData()
  const nextUserWorkplace = createWorkplace(type, id)
  updateUserData(currentUserData, { workPlace: nextUserWorkplace })
}
