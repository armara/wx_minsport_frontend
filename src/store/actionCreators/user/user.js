import flatten from 'lodash.flatten'
import api from 'utils/api'
import {
  getUserWorkplaceId,
  getWhereUserWorks as getUserWorkplaces,
} from 'utils/user'

import {
  setFacility,
  setIsLoading,
  setError,
  setAreas,
  setZones,
  setAllZones,
  setWhereUserWorks,
  setAllFacilities,
} from 'store/actions/user/user'
import { userAreasSelector, userSelectedAreaId } from 'store/selectors/user'

const facilityId = getUserWorkplaceId()

export const getFacility = () => async dispatch => {
  const response = await api().getFacilityById(facilityId)
  const { data } = response

  dispatch(setFacility(data))
}

export const getAreas = () => async dispatch => {
  dispatch(setIsLoading(true))

  let isError = false
  const result = await api()
    .getAreas()
    .catch(error => {
      isError = true
      dispatch(setError({ status: 'error', message: error }))
    })

  if (isError) return

  const { data: { areas = [] } = {} } = result
  dispatch(setAreas([...areas]))
  dispatch(setIsLoading(false))
}

export const getZones = () => async (dispatch, getState) => {
  const areaId = userSelectedAreaId(getState())
  if (!areaId) return
  dispatch(setIsLoading(true))

  let isError = false
  const result = await api()
    .getZones(areaId)
    .catch(error => {
      isError = true
      dispatch(setError({ status: 'error', message: error }))
    })
  if (isError) return

  const { data: { zones = [] } = {} } = result
  dispatch(setIsLoading(false))
  dispatch(setZones([...zones]))
}

export const getAllZones = () => async (dispatch, getState) => {
  let allAreas = userAreasSelector(getState())
  if (userAreasSelector(getState()).length === 0) {
    const { data: { areas = [] } = {} } = await api().getAreas()
    allAreas = areas
  }

  const res = await Promise.all(
    allAreas.map(async item => {
      return api().getZones(item.id)
    })
  )

  const allZones = flatten(res.map(item => item.data.zones)).map(item => ({
    title: item.name,
    ...item,
  }))

  dispatch(setAllZones(allZones))
  return allZones
}

export const getAllFacilities = () => async dispatch => {
  const facilitiesResponse = await api().getAllFacilities()
  dispatch(setAllFacilities((facilitiesResponse.data || {}).facilities || []))
}

export const getWhereUserWorks = () => async dispatch => {
  const {
    facilities = [],
    organizations = [],
    federations = [],
  } = getUserWorkplaces()

  const facilitiesResponses = await Promise.all(
    facilities.map(async id => api().getFacilityById(id))
  )
  const organizationsResponses = await Promise.all(
    organizations.map(async id => api().getOrganizationById(id))
  )
  const federationsResponses = await Promise.all(
    federations.map(async id => api().getFederationById(id))
  )

  const userFacilities = (
    facilitiesResponses.map(({ data }) => ({
      ...data,
      workplaceType: 'facilities',
    })) || []
  ).filter(workplace => !workplace.errors)
  const userOrganizations = organizationsResponses.map(
    ({ data: { data } = {} }) => ({
      ...data,
      workplaceType: 'organizations',
    })
  )
  const userFederations = federationsResponses.map(({ data }) => ({
    ...data,
    workplaceType: 'federations',
  }))

  dispatch(
    setWhereUserWorks([
      ...userFacilities,
      ...userOrganizations,
      ...userFederations,
    ])
  )
}
