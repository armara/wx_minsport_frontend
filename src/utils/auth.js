import axios from 'axios'
import debounce from 'lodash.debounce'

import { getUserData, createWorkplace, updateUserData } from 'utils/user'
import api from 'utils/api'
import history from 'utils/utils'
import dayjs, { hasTimePassed } from 'utils/day'

const MILLISECONDS_BEFORE_REFRESH_REQUEST = 1000

export const isAuthenticated = () => Object.keys(getUserData()).length > 0

const getUserDataFromResponse = response => {
  const {
    data: {
      access_token: accessToken,
      email,
      jti,
      phone,
      facilities,
      federations,
      organizations,
      ministries,
      first_name: firstName,
      sur_name: surName,
      patronymic_name: patronymicName,
      expires_in: expiresIn,
      refresh_token: refreshToken,
      token_type: tokenType,
    } = {},
  } = response

  const facilitiesIds = Object.keys(facilities)
  const federationsIds = Object.keys(federations)
  const organizationsIds = Object.keys(organizations)
  const ministriesIds = Object.keys(ministries)

  return {
    accessToken,
    email,
    jti,
    firstName: firstName || '',
    surName: surName || '',
    patronymicName: patronymicName || '',
    phone,
    facilities: facilitiesIds,
    federations: federationsIds,
    organizations: organizationsIds,
    ministries: ministriesIds,
    expiresIn,
    refreshToken,
    tokenType,
  }
}

const chooseInitialUserWorkplace = ({
  facilities,
  organizations,
  federations,
  ministries,
}) => {
  let type
  let id
  if (facilities.length > 0) {
    type = 'facilities'
    ;[id] = facilities
  } else if (organizations.length > 0) {
    type = 'organizations'
    ;[id] = organizations
  } else if (federations.length > 0) {
    type = 'federations'
    ;[id] = federations
  } else if (ministries.length > 0) {
    type = 'ministries'
    ;[id] = ministries
  }

  return createWorkplace(type, id)
}

export const login = async ({ username, password }) => {
  const formData = new FormData()
  formData.append('grant_type', 'password')
  formData.append('username', username) // petr007
  formData.append('password', password) // qwerty

  return (
    api()
      .auth(formData)
      .then(response => {
        if (response.status === 200) {
          const userData = getUserDataFromResponse(response)
          const workPlace = chooseInitialUserWorkplace(userData)
          const time = dayjs().toISOString()
          updateUserData(userData, { workPlace, time })
        }

        return response
      })
      // eslint-disable-next-line no-console
      .catch(console.error)
  )
}

export const logout = () => {
  localStorage.removeItem('user')
}

const refreshAccessToken = refreshToken => {
  const formData = new FormData()
  formData.append('grant_type', 'refresh_token')
  formData.append('refresh_token', refreshToken)

  return api()
    .auth(formData)
    .then(response => {
      if (response.status !== 200) {
        logout()
        window.location.assign('/login')
      }

      const currentUserData = getUserData()
      const nextUserData = getUserDataFromResponse(response)
      const time = dayjs().toISOString()
      updateUserData(currentUserData, { ...nextUserData, time })

      return response
    })
    .catch(() => {
      logout()
      window.location.assign('/login')
    })
}

const debouncedRefreshAccessToken = debounce(
  refreshAccessToken,
  MILLISECONDS_BEFORE_REFRESH_REQUEST,
  {
    leading: true,
    trailing: false,
  }
)

const getValidUserToken = async () => {
  const { time, expiresIn, refreshToken } = getUserData()

  const isPassed = hasTimePassed(time, expiresIn)
  if (isPassed) {
    await debouncedRefreshAccessToken(refreshToken)
    debouncedRefreshAccessToken.cancel()
  }

  return getUserData()
}

export const setupInterceptor = () => {
  axios.interceptors.response.use(
    response => {
      return response
    },
    error => {
      return Promise.reject(error)
    }
  )

  axios.interceptors.request.use(
    async config => {
      const isAuthed = isAuthenticated()

      if (config.url.endsWith('/login')) {
        return config
      }

      if (!isAuthed) {
        history.push('/login')
        return config
      }

      const { accessToken, tokenType } = await getValidUserToken()
      return {
        ...config,
        headers: {
          Authorization: `${tokenType} ${accessToken}`,
        },
      }
    },
    error => {
      return Promise.reject(error)
    }
  )
}

export const facilitiesPrefix = 'facilities'
export const organizationsPrefix = 'organizations'
