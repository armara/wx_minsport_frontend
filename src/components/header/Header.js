import React, { useEffect, useMemo, useState, useCallback } from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'
import Dotdotdot from 'react-dotdotdot'
import { connect } from 'react-redux'

import Button from 'components/button/Button'

import history from 'utils/utils'
import { logout } from 'utils/auth'
import {
  changeUserWorkplace,
  getUserPersonalFields,
  getFullUserName,
  getUserWorkplaceId,
} from 'utils/user'
import { getWhereUserWorks } from 'store/actionCreators/user/user'
import { selectWhereUserWorks } from 'store/selectors/user'

import './Header.scss'

const handleLogout = () => {
  logout()
  history.push('/login')
}

const onWorkplaceClick = (type, id) => {
  changeUserWorkplace(type, id)
  window.location.assign(`/${type}`)
}

const renderWhereUserWorks = whereUserWorks => {
  return (
    <ul>
      {whereUserWorks.map(({ name, id, workplaceType }) => (
        <li key={id}>
          <Button onClick={() => onWorkplaceClick(workplaceType, id)}>
            <>
              <i className="ic-profile-avatar" />
              <Dotdotdot tagName="div" clamp={3}>
                {name}
              </Dotdotdot>
            </>
          </Button>
        </li>
      ))}
    </ul>
  )
}

const Header = ({ whereUserWorks, fetchWhereUserWorks }) => {
  const [isWorkplacesModalOpened, setIsWorkplacesModalOpened] = useState(false)

  useEffect(() => {
    fetchWhereUserWorks()
  }, [fetchWhereUserWorks])

  const workplaceId = getUserWorkplaceId()
  const workplaceTitle = useMemo(() => {
    const workplace =
      whereUserWorks.find(({ id: _id }) => _id === workplaceId) || {}
    return workplace.name
  }, [whereUserWorks, workplaceId])

  const memoizedOnProfileClick = useCallback(() => {
    setIsWorkplacesModalOpened(!isWorkplacesModalOpened)
  }, [setIsWorkplacesModalOpened, isWorkplacesModalOpened])

  const userName = getFullUserName()
  const { email } = getUserPersonalFields()

  return (
    <header className="app-header">
      <div className="header_profile-block">
        <Button
          onClick={memoizedOnProfileClick}
          className="header_workplace-title"
          title={workplaceTitle}
        >
          <>
            {!workplaceTitle ? (
              'Загрузка...'
            ) : (
              <>
                <i className="ic-profile-avatar" />
                <div>{workplaceTitle}</div>
              </>
            )}
          </>
        </Button>
        {!workplaceTitle ? null : (
          <div
            className={classnames(
              'header_where-user-works',
              isWorkplacesModalOpened ? '' : 'hidden'
            )}
          >
            <div className="header_where-user-works_personal">
              {!userName ? null : <div>{userName}</div>}
              {!email ? null : <div>{email}</div>}
            </div>
            {renderWhereUserWorks(whereUserWorks)}
            <Button className="logout-btn" onClick={handleLogout}>
              Выйти
            </Button>
          </div>
        )}
      </div>
    </header>
  )
}

Header.propTypes = {
  whereUserWorks: PropTypes.arrayOf(PropTypes.object).isRequired,
  fetchWhereUserWorks: PropTypes.func.isRequired,
}

const mapStateToProps = state => ({
  whereUserWorks: selectWhereUserWorks(state),
})

const mapDispatchToProps = dispatch => ({
  fetchWhereUserWorks: () => dispatch(getWhereUserWorks()),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Header)
