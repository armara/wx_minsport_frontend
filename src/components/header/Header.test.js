import React from 'react'
import { configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { MemoryRouter as Router } from 'react-router-dom'
import configureStore from 'redux-mock-store'

import { updateUserData } from 'utils/user'
import { logout } from 'utils/auth'
import Header from 'components/header/Header'
import initialState, { mockUser } from 'tests/initialStoreMock'

configure({ adapter: new Adapter() })
const mockStore = configureStore()

describe('<Header />', () => {
  let store

  beforeEach(() => {
    // creates the store with any initial state or middleware needed
    store = mockStore(initialState)
    updateUserData(mockUser)
  })

  afterEach(() => {
    logout()
  })

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Router initialEntries={['/facilities']} initialIndex={0}>
          <Provider store={store}>
            <Header />
          </Provider>
        </Router>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
