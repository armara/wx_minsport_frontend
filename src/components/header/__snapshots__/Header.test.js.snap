// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> renders correctly 1`] = `
<header
  className="app-header"
>
  <div
    className="header_profile-block"
  >
    <button
      className="header_workplace-title button base-button"
      disabled={false}
      id={null}
      onClick={[Function]}
      title=""
      type="button"
    >
      
      Загрузка...
    </button>
  </div>
</header>
`;
